tools/clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -3700,7 +3700,13 @@ VisitAbstractConditionalOperator(const AbstractConditionalOperator *E) {
     return tmp5;
   }
   // HLSL Change Starts
-  if (CGF.getLangOpts().HLSL && !CGF.getLangOpts().EnableShortCircuit) {
+  // This code generates ternary results for vector and matrix types, and needs
+  // to execute if the result of the expression is a vector or matrix type even
+  // if the condition is not. Conditions are not allowed to be vector or matrix
```

Comment:
But don't we still want to avoid emitting these outside the conditional when short-circuiting?
```cpp
      llvm::Value *LHS = Visit(lhsExpr);
      llvm::Value *RHS = Visit(rhsExpr);
```
I'm not sure the fix is this simple, or I'm missing something here.

Reply:
In fact, I suspect we are missing tests for short-circuiting to ensure side-effects from dead branches are not seen for functions that return matrix and vector.

Reply:
Nope, you're right. We need to generate that under a branch instead of outside the branch.